### PR TITLE
Improvements to goto-analyzer's constant propagator and simplifier [blocks: #2522]

### PR DESCRIPTION
**Pull request by tautschnig**
Previously only `exprt::id() == ID_constant` was deemed "constant", but several other expressions (in particular: the address of an object) are also constants that can safely be propagated. The interpretation of what is (or isn't) constant now largely overlaps with what goto-symex considers a constant.

Dependencies:
- [x] #2197 
- [x] #2198 
- [x] #2199 
- [x] #2200
- [x] #2221

### CODE REVIEW

**Comment by smowton** [comment]
Looks sensible, but there ought to be corresponding new tests for the new kinds of propagation we intend to support.

File: src/util/expr_util.h [hunk 1]
```diff
@@ -72,4 +72,7 @@ bool has_subtype(const typet &, const irep_idt &id, const namespacet &);
 /// lift up an if_exprt one level
 if_exprt lift_if(const exprt &, std::size_t operand_number);
 
+/// returns true if the expression can be used as an lvalue
```
Comment (thk123) [inline src/util/expr_util.h]: As much as I hate having to add the comment, the coding standards say docs should be on the implementation.

File: src/util/replace_symbol.cpp [hunk 2]
```diff
@@ -80,7 +81,7 @@ bool replace_symbolt::replace(
     {
       const exprt &e=it->second;
 
-      if(!replace_with_const && e.is_constant())  // Would give non l-value.
+      if(!replace_with_const && !is_lvalue(e))
```
Comment (thk123) [inline src/util/replace_symbol.cpp]: This doesn't belong in this commit as this is a non-trivial change,

Previously required `replace_with_const == false` AND `e.id() == ID_constant`. Now requires `replace_with_const == false` AND a super set of the original possible `e`s.

File: src/analyses/constant_propagator.cpp [hunk 3]
```diff
@@ -278,30 +278,53 @@ bool constant_propagator_domaint::ai_simplify(
...
+  }
+  else if(expr.id() == ID_array_of)
+  {
+    return is_constant(expr.op0());
+  }
+  else if(
+    expr.id() == ID_plus || expr.id() == ID_array || expr.id() == ID_struct ||
+    expr.id() == ID_union)
+  {
+    forall_operands(it, expr)
```
Comment (thk123) [inline src/analyses/constant_propagator.cpp]: I thought `forall_operands` was now deprecated? Could use:

```C++
return std::all_of(
  expr.operands().begin(), 
  expr.operands().end(), 
  [](const exprt &) { return is_constant(expr); });
```
To make the condition more explicit

File: src/analyses/constant_propagator.cpp [hunk 4]
```diff
@@ -278,30 +278,53 @@ bool constant_propagator_domaint::ai_simplify(
...
 
-  if(expr.id()==ID_index)
-    return false;
+    return is_constant(typecast_expr.op());
+  }
+  else if(expr.id() == ID_array_of)
+  {
+    return is_constant(expr.op0());
+  }
+  else if(
```
Comment (thk123) [inline src/analyses/constant_propagator.cpp]: ID_minus, ID_multiply etc? Or am I missing something

**Review by thk123 — COMMENTED** [review]
Agree with @smowton this needs some tests - also dunno how much overlap there is with this + variable sensitivity domain enhancements to constant propagation @chrisr-diffblue

**Comment by tautschnig** [comment]
Tests have now been added - without adding any new tests...! I have gone through various regression tests and moved from `FUTURE` to `CORE` those tests that the individual commits now render successful.

**Review by smowton — APPROVED** [review]

File: regression/goto-analyzer/constant_propagation_09/test.desc [hunk 5]
```diff
@@ -3,7 +3,7 @@ main.c
 --constants --simplify out.gb
 ^EXIT=0$
 ^SIGNAL=0$
-^Simplified:  assert: 1, assume: 0, goto: 3, assigns: 4, function calls: 0$
-^Unmodified:  assert: 0, assume: 0, goto: 0, assigns: 11, function calls: 2$
+^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 6, function calls: 0$
+^Unmodified:  assert: 0, assume: 0, goto: 1, assigns: 9, function calls: 2$
 --
```
Comment (martin-cs) [inline regression/goto-analyzer/constant_propagation_09/test.desc:7]: I'm not sure this change matches the  commit message of the commit it is in but I guess it is being updated elsewhere in the patch set, in which case it's a bit unnecessary to change it.

File: src/util/replace_symbol.cpp (hunk 2 above)
Comment (martin-cs) [reply src/util/replace_symbol.cpp]: Like @thk123 I'm reviewing this patch by patch and yes, there is an issue here.  I'm happy to resolve it by seeing where else in the patch set it would fit rather than moving it but still.

File: src/analyses/constant_propagator.cpp [hunk 6]
```diff
@@ -278,30 +278,53 @@ bool constant_propagator_domaint::ai_simplify(
 
 bool constant_propagator_domaint::valuest::is_constant(const exprt &expr) const
 {
-  if(expr.id()==ID_side_effect &&
-     to_side_effect_expr(expr).get_statement()==ID_nondet)
-    return false;
+  if(expr.is_constant())
+    return true;
```
Comment (martin-cs) [inline src/analyses/constant_propagator.cpp]: Is there any way we could share the code between this and symex rather than keeping the two maintained in parallel? expr_util.h?

File: src/goto-analyzer/static_simplifier.h [hunk 7]
```diff
@@ -18,7 +18,7 @@ class optionst;
 
 bool static_simplifier(
   goto_modelt &,
-  const ai_baset &,
+  ai_baset &,
   const optionst &,
```
Comment (martin-cs) [inline src/goto-analyzer/static_simplifier.h]: Yeah... I think we need to talk about this one.

File: src/analyses/constant_propagator.cpp [hunk 8]
```diff
@@ -269,10 +269,18 @@ bool constant_propagator_domaint::ai_simplify(
   exprt &condition,
   const namespacet &ns) const
 {
-  bool b=values.replace_const.replace(condition);
-  b&=simplify(condition, ns);
+  bool no_change = true;
 
-  return b;
+  while(!values.replace_const.replace(condition))
+  {
```
Comment (martin-cs) [inline src/analyses/constant_propagator.cpp]: Is there a test case for this?  As in; do you know where it is needed?

File: src/goto-programs/generate_function_bodies.h [hunk 9]
```diff
@@ -74,11 +74,13 @@ void generate_function_bodies(
...
+  "                              Generate bodies for functions "               \
+  "matching regex\n"                                                           \
   " --generate-function-body-options <option>\n"                               \
   "                              One of assert-false, assume-false,\n"         \
-  "                              nondet-return, assert-false-assume-false and" \
-  "                              \nhavoc[,params:<regex>][,globals:<regex>]\n" \
+  "                              nondet-return, assert-false-assume-false "    \
+  "and\n"                                                                      \
+  "                              havoc[,params:<regex>][,globals:<regex>]\n"   \
   "                              (default: nondet-return)"
```
Comment (martin-cs) [inline src/goto-programs/generate_function_bodies.h]: Unrelated, but, yes.

File: src/goto-analyzer/static_simplifier.cpp [hunk 10]
```diff
@@ -220,6 +221,10 @@ bool static_simplifier(
              << ", function calls: " << unmodified.function_calls
              << messaget::eom;
 
+  // restore return types before writing the binary
+  restore_returns(goto_model);
+  goto_model.goto_functions.update();
```
Comment (martin-cs) [inline src/goto-analyzer/static_simplifier.cpp]: Sure.  I know I keep saying this BUT we (in the broadest sense) do need to work out what the valid forms of goto-program are at various points in the tool.  This reads to me a little bit like "returns must always be present when on disk".

File: src/goto-programs/remove_returns.cpp [hunk 11]
```diff
@@ -352,38 +352,9 @@ bool remove_returnst::restore_returns(
...
-      }
-      else
-      {
-        INVARIANT(
-          i_it->is_end_function(),
-          "control-flow after assigning return value should lead directly "
-          "to end of function");
-        i_it=goto_program.instructions.insert(i_it, *i_it);
-      }
-
```
Comment (martin-cs) [inline src/goto-programs/remove_returns.cpp]: Again; this looks like another "we need to clarify our invariants" issue.

**Review by martin-cs — CHANGES_REQUESTED** [review]
A few changes but more than that, I'd like to talk about the patch that loops calls to AI.

**Comment by martin-cs** [comment]
I have two things against the looped calls to AI patch (the rest I'm happy to accept (although having the constant detection shared rather than duplicated would be awesome)):

1. It rather breaks the idea of goto-analyze as "run AI; then do task".  That's not the end of the world but it would be nice and more specifically I'd want a good reason for it.  Which brings me to...

2. Isn't this just papering over the cracks?  I think simplify should be idempotent.  I think it not being is indicative of a bug / lack of precision.  @hannes-steffenhagen-diffblue very kindly did some work on this with the variable sensitivity domain and found a number of issues.  So I'm kind of inclined to look at the cases where repeated simplification helps and see if they are actually issues with the domain / simplification code.

**Comment by martin-cs** [comment]
( I realise this week I am "block and discuss" on every patch but ... )

**Comment by tautschnig** [comment]
> ( I realise this week I am "block and discuss" on every patch but ... )

I do strongly prefer to have good quality code so all those inputs are highly appreciated! So ...

> 2. Isn't this just papering over the cracks?  I think simplify should be idempotent.  I think it not being is indicative of a bug / lack of precision.  @hannes-steffenhagen-diffblue very kindly did some work on this with the variable sensitivity domain and found a number of issues.  So I'm kind of inclined to look at the cases where repeated simplification helps and see if they are actually issues with the domain / simplification code.

I don't think it's as simple as that; in fact simplify, as far as I can tell, _is_ idempotent. There might be some bugs left, but that's not what this looping is about. The key bit is re-running constant propagation (or indeed whatever domain) after having done simplification. Consider a reduced version of one of the regression tests:
```
int i = 0;
if(i == 0) ++i;
assert(i == 1);
```
The first iteration of constant propagation will determine the value of `i` for all instructions in line 2, but cannot resolve the assertion, because there still is a join point leaving the value of `i` non-constant at that point. Thus the result of the first run is
```
int i = 0;
i = 1;
assert(i == 1);
```
Now the branching is gone, and constant propagation plus simplification can happily sort out the assertion. I am all ears to hear about better solutions, but from my point of view a fixed point around the combination of abstraction interpretation and simplification is required here.

File: src/goto-analyzer/static_simplifier.cpp (hunk 10 above)
Comment (tautschnig) [reply src/goto-analyzer/static_simplifier.cpp]: It's not as much about returns, but more about the types: I don't think the simplifier should yield a binary where symbols have different types. Or at least that's not what I'd expect from a "simplifier".

File: src/goto-programs/remove_returns.cpp (hunk 11 above)
Comment (tautschnig) [reply src/goto-programs/remove_returns.cpp]: Agreed, though in this case that was code I had originally corrupted^W^Wwritten with fixed assumptions about its structure. Those assumptions were unnecessary.

File: src/goto-analyzer/static_simplifier.cpp (hunk 10 above)
Comment (martin-cs) [reply src/goto-analyzer/static_simplifier.cpp]: Fair.

File: src/goto-programs/remove_returns.cpp (hunk 11 above)
Comment (martin-cs) [reply src/goto-programs/remove_returns.cpp]: Also fair.

File: src/analyses/constant_propagator.cpp (hunk 8 above)
Comment (tautschnig) [reply src/analyses/constant_propagator.cpp]: That's needed when local constant propagation yields a subexpression that is not an lvalue, as in the following:
```
int x = 1;
int *p = &x;
x = *p;
```
Now the first substitution+simplification will be `x = x;` when only the second round will yield `x = 1;`.

File: src/analyses/constant_propagator.cpp (hunk 6 above)
Comment (tautschnig) [reply src/analyses/constant_propagator.cpp]: I'll give this a try, even though symex has some peculiarities around plus and multiplication just so as to feed more information to calls to `malloc`.

**Comment by martin-cs** [comment]
> > 2. Isn't this just papering over the cracks?  I think simplify should be idempotent.  I think it not being is indicative of a bug / lack of precision.  @hannes-steffenhagen-diffblue very kindly did some work on this with the variable sensitivity domain and found a number of issues.  So I'm kind of inclined to look at the cases where repeated simplification helps and see if they are actually issues with the domain / simplification code.
> 
> I don't think it's as simple as that; in fact simplify, as far as I can tell, _is_ idempotent. There might be some bugs left, but that's not what this looping is about. The key bit is re-running constant propagation (or indeed whatever domain) after having done simplification.

I was including this in the idempotency part!

>  Consider a reduced version of one of the regression tests:
> ```
> int i = 0;
> if(i == 0) ++i;
> assert(i == 1);
> ```
> The first iteration of constant propagation will determine the value of `i` for all instructions in line 2, but cannot resolve the assertion, because there still is a join point leaving the value of `i` non-constant at that point.

I see that as a precision issue.  I think only one branch of the if
statement should be taken because ```i == 0``` can be resolved by the
domain.  Thus one side will be \bot and so the assertion can be reduced
to ```assert(true)```.

>  Thus the result of the first run is
> ```
> int i = 0;
> i = 1;
> assert(i == 1);
> ```
> Now the branching is gone, and constant propagation plus simplification can happily sort out the assertion. I am all ears to hear about better solutions, but from my point of view a fixed point around the combination of abstraction interpretation and simplification is required here.

Do you have examples of this causing a problem?  My next real
development task is finishing off goto-analyze and the variable
sensitivity domain.  It is my belief that all cases of needing multiple
iterations I can fix with the VSD.  Let's give it a go and see if we can
solve this by finding real programs that have this issue.

**Comment by martin-cs** [comment]
> That's needed when local constant propagation yields a subexpression that is not an lvalue, as in the following:
> ```
> int x = 1;
> int *p = &x;
> x = *p;
> ```
> Now the first substitution+simplification will be `x = x;` when only the second round will yield `x = 1;`.

Again, in my eyes, this looks like a precision problem where there is a
lack of handling of pointers (a design problem in the constant
propagator, which is why I think we should eventually deprecate it in
favour of VSD).

**Comment by tautschnig** [comment]
[...]
> >  Consider a reduced version of one of the regression tests:
> > ```
> > int i = 0;
> > if(i == 0) ++i;
> > assert(i == 1);
> > ```
> > The first iteration of constant propagation will determine the value of `i` for all instructions in line 2, but cannot resolve the assertion, because there still is a join point leaving the value of `i` non-constant at that point.

> I see that as a precision issue.  I think only one branch of the if
statement should be taken because ```i == 0``` can be resolved by the
domain.  Thus one side will be \bot and so the assertion can be reduced
to ```assert(true)```.

Ok, but that's not what the domain is doing right now? So it's a but in the constant-propagation domain?

> >  Thus the result of the first run is
> > ```
> > int i = 0;
> > i = 1;
> > assert(i == 1);
> > ```
> > Now the branching is gone, and constant propagation plus simplification can happily sort out the assertion. I am all ears to hear about better solutions, but from my point of view a fixed point around the combination of abstraction interpretation and simplification is required here.

> Do you have examples of this causing a problem?  My next real
development task is finishing off goto-analyze and the variable
sensitivity domain.  It is my belief that all cases of needing multiple
iterations I can fix with the VSD.  Let's give it a go and see if we can
solve this by finding real programs that have this issue.

Yes, I do (as otherwise admittedly wouldn't have touched this code...). My workflow includes using the static unwinder that @danpoe built before applying the simplifier. I'm all ears if there is a better solution. I might actually filter out the undisputed bug fixes from this PR into a new one so that those can go in in the meantime.

**Comment by tautschnig** [comment]
I have started breaking out smaller fixes in #2197, #2198, #2199 (and at least one more to come).

**Comment by martin-cs** [comment]
On Thu, 2018-05-17 at 13:13 -0700, Michael Tautschnig wrote:
> I have started breaking out smaller fixes in #2197, #2198, #2199 (and at least one more to come).

Cool; thank you.

File: src/util/replace_symbol.cpp (hunk 2 above)
Comment (tautschnig) [reply src/util/replace_symbol.cpp]: This is a separate PR now: #2200.

**Comment by tautschnig** [comment]
This is now rebased and ready for re-review - @martin-cs in particular.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4ea7de3b3c661507f0d72c664b862445a1a82e19).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90620225
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by tautschnig** [comment]
Pinging @martin-cs, @chrisr-diffblue, @danpoe, @smowton as I added some new commits in here. Feedback much appreciated.

File: src/analyses/constant_propagator.cpp [hunk 12]
```diff
@@ -589,11 +602,21 @@ bool constant_propagator_domaint::partial_evaluate_with_all_rounding_modes(
...
+  const valuest &known_values,
   exprt &expr,
-  const namespacet &ns) const
+  const namespacet &ns)
 {
-  bool did_not_change_anything = values.replace_const.replace(expr);
-  did_not_change_anything &= simplify(expr, ns);
+  bool did_not_change_anything = true;
+
+  while(!known_values.replace_const.replace(expr))
```
Comment (chrisr-diffblue) [inline src/analyses/constant_propagator.cpp:622]: While I think I understand the purpose of the while loop here, and the purpose of the final `if(...) simplify()` following the while loop, it's not entirely obvious so might be worth a comment or two to explain?

File: regression/goto-analyzer/constant_propagation_19/test.desc [hunk 13]
```diff
@@ -0,0 +1,8 @@
+CORE
```
Comment (chrisr-diffblue) [inline regression/goto-analyzer/constant_propagation_19/test.desc:1]: Would it be possible to call this new test something a little more descriptive, such as `constant_propagation_two_way_1` or something?

Comment (tautschnig) [reply regression/goto-analyzer/constant_propagation_19/test.desc:1]: Done.

File: src/analyses/constant_propagator.cpp (hunk 12 above)
Comment (tautschnig) [reply src/analyses/constant_propagator.cpp:622]: Done.

File: src/analyses/constant_propagator.cpp [hunk 14]
```diff
@@ -36,26 +36,50 @@ void constant_propagator_domaint::assign_rec(
...
+        base_type_eq(ns.lookup(s).type, tmp.type(), ns),
+        "type of constant to be replaced should match");
+      dest_values.set_to(s, tmp);
+    }
+    else
+      dest_values.set_to_top(s);
   }
   else
-    dest_values.set_to_top(s);
+    dest_values.set_to_top();
```
Comment (chrisr-diffblue) [inline src/analyses/constant_propagator.cpp:92]: Worth adding a comment to explain that this is for soundness?

**Review by chrisr-diffblue — APPROVED** [review]
I'm not as averse to the looping as @Martin-cs , but I also admit I don't have his experience either :-) I think changes in this PR have made the code cleaner and clearly better able to propagate than it does right now, so I'm approving on those basis.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d07507bd35c68aa445ab9518896187f64efa7247).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91427232
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 03dfcb00c05d1ce940b90574edd4855f9034ce38).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91431166

File: src/analyses/constant_propagator.cpp [hunk 15]
```diff
@@ -566,12 +572,12 @@ bool constant_propagator_domaint::partial_evaluate_with_all_rounding_modes(
   exprt first_result;
   for(std::size_t i = 0; i < rounding_modes.size(); ++i)
   {
-    constant_propagator_domaint child(*this);
-    child.values.set_to(
+    valuest beacon_values = known_values;
```
Comment (smowton) [inline src/analyses/constant_propagator.cpp]: Beacon? Haven't heard that term before

File: src/analyses/constant_propagator.cpp [hunk 16]
```diff
@@ -623,34 +630,34 @@ void constant_propagator_ait::replace(
 
     if(it->is_goto() || it->is_assume() || it->is_assert())
     {
-      d.partial_evaluate(it->guard, ns);
+      constant_propagator_domaint::partial_evaluate(d.values, it->guard, ns);
```
Comment (smowton) [inline src/analyses/constant_propagator.cpp:662]: I don't think you need the `constant_propagator_domaint::` qualifier since this method lives in the same namespace

**Review by smowton — APPROVED** [review]
Looks good to me -- any thoughts on how this interacts with the improvements in https://github.com/diffblue/cbmc/pull/2522 ?

File: src/analyses/constant_propagator.cpp (hunk 15 above)
Comment (tautschnig) [reply src/analyses/constant_propagator.cpp]: Should I just use `tmp_values`? I'll go with that for the moment, all ideas appreciated.

File: src/analyses/constant_propagator.cpp (hunk 16 above)
Comment (tautschnig) [reply src/analyses/constant_propagator.cpp:662]: I don't think so - it's here being invoked from `constant_propagator_ait::replace` (note that I've made `partial_evaluate` a static member of `constant_propagator_domaint`).

**Comment by tautschnig** [comment]
> any thoughts on how this interacts with the improvements in #2522 ?

Let me comment on #2522, many thanks for refreshing my memory on that.

File: src/analyses/constant_propagator.cpp (hunk 16 above)
Comment (smowton) [reply src/analyses/constant_propagator.cpp:662]: Ah sorry, misread

File: src/analyses/constant_propagator.cpp (hunk 15 above)
Comment (smowton) [reply src/analyses/constant_propagator.cpp]: Either that or comment to say what a beacon is in this context :) I assume that's a term used elsewhere, I just haven't heard of it

Comment (tautschnig) [reply src/analyses/constant_propagator.cpp]: That was just me making things up with my limited command of the language.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9b16c48f93e5c2bab22ac3ded79ad4f828b6f743).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91505755

File: regression/goto-analyzer/sensitivity-function-call-recursive/test.desc [hunk 17]
```diff
@@ -1,4 +1,4 @@
-KNOWNBUG
+FUTURE
 main.c
 --variable --pointers --arrays --structs --verify
```
Comment (martin-cs) [inline regression/goto-analyzer/sensitivity-function-call-recursive/test.desc:3]: These should be future, also, the names for options should change.

File: regression/goto-analyzer/constant_propagation_09/test.desc [hunk 18]
```diff
@@ -1,9 +1,8 @@
-FUTURE
+KNOWNBUG
 main.c
---constants --simplify out.gb
+--constants --verify
```
Comment (martin-cs) [inline regression/goto-analyzer/constant_propagation_09/test.desc:3]: Why has this changed task?

File: regression/goto-analyzer/constant_propagation_11/test.desc [hunk 19]
```diff
@@ -1,9 +1,8 @@
-FUTURE
+KNOWNBUG
 main.c
---constants --simplify out.gb
+--constants --verify
```
Comment (martin-cs) [inline regression/goto-analyzer/constant_propagation_11/test.desc:3]: Again, why change task?

File: regression/goto-analyzer/constant_propagation_14/test.desc [hunk 20]
```diff
@@ -1,9 +1,9 @@
-FUTURE
+KNOWNBUG
```
Comment (martin-cs) [inline regression/goto-analyzer/constant_propagation_14/test.desc]: This change, although not wrong, doesn't seem to match the commit message.

**Review by martin-cs — APPROVED** [review]
Yes; looks much better.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 731c3bf292d9fd28061cfcdfa05a59c8b1dce74b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93273597

File: regression/goto-analyzer/constant_propagation_09/test.desc (hunk 18 above)
Comment (tautschnig) [reply regression/goto-analyzer/constant_propagation_09/test.desc:3]: For completeness: I have changed the task here as simplification did always take place, but not enough of it to actually turn the assertion into a firm `FALSE`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 23596895e341d41fca4068645a8ddf2abf18ccce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95302081
